### PR TITLE
v4.0.x: Fixed uninitialzed memory access bug in base64 encoding.

### PR DESCRIPTION
**Pull request by hppritcha**
Signed-off-by: Charles Shereda <cpshereda@lanl.gov>
(cherry picked from commit 97b28732e4ab516cd5c273e0d60d124a537f51b4)

### CODE REVIEW

File: opal/mca/pmix/base/pmix_base_fns.c [hunk 1]
```diff
@@ -663,9 +665,11 @@ static inline unsigned char pmi_base64_decsym (unsigned char value) {
 
 static inline void pmi_base64_encode_block (const unsigned char in[3], char out[4], int len) {
     out[0] = pmi_base64_encsym (in[0] >> 2);
-    out[1] = pmi_base64_encsym (((in[0] & 0x03) << 4) | ((in[1] & 0xf0) >> 4));
+    /* len == length of in[] - conditionals insure we don't reference uninitialized in[] values */
+    out[1] = 1 < len ? pmi_base64_encsym(((in[0] & 0x03) << 4) | ((in[1] & 0xf0) >> 4)) : pmi_base64_encsym((in[0] & 0x03) << 4);
     /* Cray PMI doesn't allow = in PMI attributes so pad with spaces */
-    out[2] = 1 < len ? pmi_base64_encsym(((in[1] & 0x0f) << 2) | ((in[2] & 0xc0) >> 6)) : ' ';
+    out[2] = 1 < len ? pmi_base64_encsym((in[1] & 0x0f) << 2) : ' ';
+    out[2] = 2 < len ? pmi_base64_encsym(((in[1] & 0x0f) << 2) | ((in[2] & 0xc0) >> 6)) : out[2];
```
Comment (gpaulsen) [inline opal/mca/pmix/base/pmix_base_fns.c:672]: Is it intended to write to `out[2]` twice like this?

**Review by gpaulsen — CHANGES_REQUESTED** [review]
Sorry, I just glanced at the code.

How is this related to uninitialized mem?

File: opal/mca/pmix/base/pmix_base_fns.c (hunk 1 above)
Comment (hppritcha) [reply opal/mca/pmix/base/pmix_base_fns.c:672]: @cpshereda does this patch look correct?  I just cherry-picked from #8265

Comment (cpshereda) [reply opal/mca/pmix/base/pmix_base_fns.c:672]: @gpaulsen @hppritcha  Re: writing out[2] twice: Yes, it's written twice if len > 2. It was a way to preserve the original style while fixing the bug and it doesn't hurt anything AFAIK (the problem was originally discovered in PMIx and is not performance-dependent, if you're worried about the two assignments for performance reasons, that's a different matter)

Re: how this is related to uninitialized mem: `len` is the length of `in[]`. In the original, we are potentially checking `in[2]` in cases where `len` is only equal to 2.

Comment (cpshereda) [reply opal/mca/pmix/base/pmix_base_fns.c:672]: @hppritcha @gpaulsen To be clear, the same type of issue is present in the two changed assignments.

Howard: Yes, that looks like the original fix.

**Review by gpaulsen — APPROVED** [review]
Thanks.  Looks good, and consistent with v4.1 commit.

**Pull request merged**